### PR TITLE
Mattermost container

### PR DESCRIPTION
**Pull request by snaselj**
# Partial Work: #194

## What's Changed

- Prepared for integrations.
- Added Mattermost container for the development.
- Improved invoke tasks.

## TODO

- [x] Bump minimal Nautobot version to 1.5.2.
- [x] Use environment variables to enable Platforms.
- [x] Make Nautobot with Mattermost bootstrap working.
- [x] Improve docs (glossary, quickstart, bootstrap).

### CODE REVIEW

File: development/nautobot_config.py [hunk 1]
```diff
@@ -117,26 +117,30 @@
...
+    "nautobot_chatops",
+    # "nautobot_chatops.tests.bootstrap",
+    "nautobot_capacity_metrics",
+]
 
 # Plugins configuration settings. These settings are used by various plugins that the user may have installed.
 # Each key in the dictionary is the name of an installed plugin and its value is a dictionary of settings.
 PLUGINS_CONFIG = {
     "nautobot_chatops": {
-        "enable_slack": True,
```
Comment (smk4664) [inline development/nautobot_config.py:126]: These should change to be environment variables, and default to false. Then we won't need to comment them out.

**Review by smk4664 — COMMENTED** [review]
I will review this PR in a bit, but I left one comment for now.

File: development/nautobot_config.py (hunk 1 above)
Comment (snaselj) [reply development/nautobot_config.py:126]: Fixed

**Comment by snaselj** [comment]
@smk4664 Code should be ready for review now, just a question regarding docs. How would you organize changes to README.md? Should I rather put quickstart to `docs/dev`?

File: tasks.py [hunk 2]
```diff
@@ -355,6 +384,8 @@ def build_and_check_docs(context):
 )
 def unittest(context, keepdb=False, label="nautobot_chatops", failfast=False, buffer=True):
     """Run Nautobot unit tests."""
+    docker_compose(context, "rm --stop -- nautobot")
```
Comment (bryanculver) [inline tasks.py:387]: Let's be sure to add a comment about why this is in here.

File: tasks.py [hunk 3]
```diff
@@ -150,6 +152,12 @@ def generate_packages(context):
 # ------------------------------------------------------------------------------
 # START / STOP / DEBUG
 # ------------------------------------------------------------------------------
+@task
+def export(context):
+    """Export docker compose configuration to `compose.yaml` file."""
+    docker_compose(context, "convert > compose.yaml")
```
Comment (bryanculver) [inline tasks.py:158]: Should we upstream this? Can we describe when you may want to use the exported files?

File: development/docker-compose.dev.yml [hunk 4]
```diff
@@ -24,3 +24,19 @@ services:
...
+      MM_SERVICESETTINGS_SITEURL: "http://mattermost:8065"
+      MM_TEAMSETTINGS_ENABLEOPENSERVER: "true"
+    ports:
+      - "8065:8065"
+    volumes:
+      - "./mattermost/dump.sql:/docker-entrypoint-initdb.d/dump.sql"
+      - "mattermost-data:/var/lib/mysql"
+
+volumes:
+  mattermost-data: {}
```
Comment (bryanculver) [inline development/docker-compose.dev.yml:42]: Let's move this out to its own compose file and mimic what Core uses for SSO: https://github.com/nautobot/nautobot/blob/develop/development/docker-compose.keycloak.yml

This allows adding a single line into `invoke.yml`: https://github.com/nautobot/nautobot/blob/bdf1faa6386a0fa866978f6eb866413a336cb9f9/invoke.yml.example#L37

Then the environment variable set in `dev.env` can be automatically applied to the right containers.

File: tasks.py (hunk 2 above)
Comment (snaselj) [reply tasks.py]: This line was removed as I removed Bootstrap plugin. With the current Nautobot implementation, it's not possible to have multiple Apps in a single package: https://github.com/nautobot/nautobot/blob/bdf1faa6386a0fa866978f6eb866413a336cb9f9/nautobot/extras/plugins/urls.py#L27

File: tasks.py (hunk 3 above)
Comment (snaselj) [reply tasks.py]: Improved docs, should I add it to cookiecutter templates and nautobot repository as well?

File: development/docker-compose.dev.yml (hunk 4 above)
Comment (snaselj) [reply development/docker-compose.dev.yml]: Applied.

I moved mattermost compose configuration to `development/mattermost/docker-compose.yml`, WDYT @bryanculver ?

**Comment by snaselj** [comment]
This PR is ready for review now, some docs improvements is still necessary.

**Comment by whitej6** [comment]
If we are bumping minimum Nautobot version to the 1.5 train we should bump to 1.5.4 I believe which has a more robust ui viewset and other features in the apps api

File: nautobot_chatops/dispatchers/__init__.py [hunk 5]
```diff
@@ -3,7 +3,3 @@
 The classes implemented in this module implement a common API for "dispatching" content
 (plaintext messages, formatted text, UI elements, etc.) to the user of a given chat platform.
 """
-
-from .base import Dispatcher
-
-__all__ = ("Dispatcher",)
```
Comment (bryanculver) [inline nautobot_chatops/dispatchers/__init__.py:9]: Why was this needed to add a dev container?

File: README.md [hunk 6]
```diff
@@ -47,3 +47,30 @@ Interested to see Nautobot ChatOps in action?  It's currently setup on the [Demo
...
+git clone git@github.com:nautobot/nautobot-plugin-chatops.git
+cd nautobot-plugin-chatops
+inv build
+inv start
+# Nautobot available as http://127.0.0.1:8080 admin / admin
+# Mattermost available at http://127.0.0.1:8065 admin / admin
+
+# To allow Mattermost integration run the following after Nautobot starts:
+inv bootstrap-mattermost
+```
```
[the remark this answers is not in the document]
Comment (smk4664) [reply README.md:65]: Agreed, and I believe @snaselj  that this answers one of your questions for me.

File: nautobot_chatops/utils.py [hunk 7]
```diff
@@ -33,10 +33,7 @@ def celery_worker_task(command, subcommand, params, dispatcher_module, dispatche
         # Looking up the function from the registry using the command.
         registry = get_commands_registry()
         function = registry[command]["function"]
-        # Get dispatcher class from module, since the module is already loaded, we can use
-        # sys.modules to map the dispatcher module string to the module. Then pull the class
-        # using getattr.
-        dispatcher_class = getattr(sys.modules[dispatcher_module], dispatcher_name)
+        dispatcher_class = getattr(import_module(dispatcher_module), dispatcher_name)
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_chatops/utils.py:36]: Yes, the original version didn't work for me, this is a bugfix. My guess is that was working with RQ, but not with Celery.

File: nautobot_chatops/dispatchers/__init__.py (hunk 5 above)
Comment (snaselj) [reply nautobot_chatops/dispatchers/__init__.py]: Fixed, reverted

File: nautobot_chatops/__init__.py [hunk 8]
```diff
@@ -7,10 +7,10 @@
 
 __version__ = metadata.version(__name__)
 
-from nautobot.extras.plugins import PluginConfig
+from nautobot.apps import NautobotAppConfig
 
 
-class NautobotChatOpsConfig(PluginConfig):
+class NautobotChatOpsConfig(NautobotAppConfig):
```
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_chatops/__init__.py]: Fixed, reverted

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (snaselj) [reply README.md]: Done

**Comment by snaselj** [comment]
All comments are addressed, PR can be reviewed.

File: docs/dev/design.md [hunk 9]
```diff
@@ -5,6 +5,16 @@ By delivering this as a Nautobot ChatOps plugin, we gain the following benefits:
 - No need to stand up a separate HTTP server, just use Nautobot's own HTTP server.
 - Use of `async` is not required because we can hand off long-running tasks to Nautobot's `celery` workers. With one notable exception, WebSocket connections will be done with `async`. We have opted for `AIOHTTP` for our WebSocket [clients](https://docs.aiohttp.org/en/stable/client_quickstart.html#aiohttp-client-websockets).
 
+## Glossary
+
+- Platform: A communication service, such as Slack, Microsoft Teams, or Mattermost.
+- Dispatcher: A Platform specific class that processes incoming chat messages, executes Commands, and sends results back to the Platform.
+- Platform View: Django View to handle Platform data.
+- Provider: A component that defines Commands, handles data retrieval, and manipulation for particular tools or services (e.g., IP Fabric, Ansible, Panorama, Grafana, ACI, Meraki, Arista CloudVision).
```
Comment (smk4664) [inline docs/dev/design.md:13]:
```suggestion
- Controller: A component that defines Commands, handles data retrieval, and manipulation for particular tools or services (e.g., IP Fabric, Ansible, Panorama, Grafana, ACI, Meraki, Arista CloudVision).
```

File: docs/dev/design.md [hunk 10]
```diff
@@ -5,6 +5,16 @@ By delivering this as a Nautobot ChatOps plugin, we gain the following benefits:
...
+- Command: A specific instruction sent by users through the chat Platform, processed by some Provider.
```
Comment (smk4664) [inline docs/dev/design.md:14]:
```suggestion
- Command: A specific instruction sent by users through the chat Platform, processed by some Controller.
```

File: nautobot_chatops/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (smk4664) [reply nautobot_chatops/utils.py:36]: Interesting... I have been using these just fine on Celery for the last year.

**Review by smk4664 — COMMENTED** [review]
Thanks for the changes, I have a few more suggestions, but otherwise this looks great!

File: nautobot_chatops/utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (snaselj) [reply nautobot_chatops/utils.py]: My bad, I was missing Mattermost configuration for Celery worker, fixed and reverted .

File: docs/dev/design.md (hunk 9 above)
Comment (snaselj) [reply docs/dev/design.md]: Provider renamed to Integration as specified as @bryanculver

File: docs/dev/design.md (hunk 10 above)
Comment (snaselj) [reply docs/dev/design.md]: Provider renamed to Integration as specified as @bryanculver

**Comment by snaselj** [comment]
Ready for review, all comments are addressed @bryanculver @smk4664

**Review by smk4664 — APPROVED** [review]
Thanks Jan for making all these changes! It will be great being able to spin up Mattermost quickly for development.
